# 02_Spark_Dataframe

#### Ref:
1. Spark official link: https://spark.apache.org/
2. Spark Python API: https://spark.apache.org/docs/latest/api/python/reference/index.html

#### How to list all available datasets from databricks cluster

In [0]:
%fs ls /databricks-datasets

path name size modificationTime dbfs:/databricks-datasets/ databricks-datasets/ 0 0 dbfs:/databricks-datasets/COVID/ COVID/ 0 0 dbfs:/databricks-datasets/README.md README.md 976 1532468253000 dbfs:/databricks-datasets/Rdatasets/ Rdatasets/ 0 0 dbfs:/databricks-datasets/SPARK_README.md SPARK_README.md 3359 1455043490000 dbfs:/databricks-datasets/adult/ adult/ 0 0 dbfs:/databricks-datasets/airlines/ airlines/ 0 0 dbfs:/databricks-datasets/amazon/ amazon/ 0 0 dbfs:/databricks-datasets/asa/ asa/ 0 0 dbfs:/databricks-datasets/atlas_higgs/ atlas_higgs/ 0 0 dbfs:/databricks-datasets/bikeSharing/ bikeSharing/ 0 0 dbfs:/databricks-datasets/cctvVideos/ cctvVideos/ 0 0 dbfs:/databricks-datasets/credit-card-fraud/ credit-card-fraud/ 0 0 dbfs:/databricks-datasets/cs100/ cs100/ 0 0 dbfs:/databricks-datasets/cs110x/ cs110x/ 0 0 dbfs:/databricks-datasets/cs190/ cs190/ 0 0 dbfs:/databricks-datasets/data.gov/ data.gov/ 0 0 dbfs:/databricks-datasets/definitive-guide/ definitive-guide/ 0 0 dbfs:/databricks-datasets/delta-sharing/ delta-sharing/ 0 0 dbfs:/databricks-datasets/flights/ flights/ 0 0 dbfs:/databricks-datasets/flower_photos/ flower_photos/ 0 0 dbfs:/databricks-datasets/flowers/ flowers/ 0 0 dbfs:/databricks-datasets/genomics/ genomics/ 0 0 dbfs:/databricks-datasets/hail/ hail/ 0 0 dbfs:/databricks-datasets/identifying-campaign-effectiveness/ identifying-campaign-effectiveness/ 0 0 dbfs:/databricks-datasets/iot/ iot/ 0 0 dbfs:/databricks-datasets/iot-stream/ iot-stream/ 0 0 dbfs:/databricks-datasets/learning-spark/ learning-spark/ 0 0 dbfs:/databricks-datasets/learning-spark-v2/ learning-spark-v2/ 0 0 dbfs:/databricks-datasets/lending-club-loan-stats/ lending-club-loan-stats/ 0 0 dbfs:/databricks-datasets/med-images/ med-images/ 0 0 dbfs:/databricks-datasets/media/ media/ 0 0 dbfs:/databricks-datasets/mnist-digits/ mnist-digits/ 0 0 dbfs:/databricks-datasets/news20.binary/ news20.binary/ 0 0 dbfs:/databricks-datasets/nyctaxi/ nyctaxi/ 0 0 dbfs:/databricks-datasets/nyctaxi-with-zipcodes/ nyctaxi-with-zipcodes/ 0 0 dbfs:/databricks-datasets/online_retail/ online_retail/ 0 0 dbfs:/databricks-datasets/overlap-join/ overlap-join/ 0 0 dbfs:/databricks-datasets/power-plant/ power-plant/ 0 0 dbfs:/databricks-datasets/retail-org/ retail-org/ 0 0 dbfs:/databricks-datasets/rwe/ rwe/ 0 0 dbfs:/databricks-datasets/sai-summit-2019-sf/ sai-summit-2019-sf/ 0 0 dbfs:/databricks-datasets/sample_logs/ sample_logs/ 0 0 dbfs:/databricks-datasets/samples/ samples/ 0 0 dbfs:/databricks-datasets/sfo_customer_survey/ sfo_customer_survey/ 0 0 dbfs:/databricks-datasets/sms_spam_collection/ sms_spam_collection/ 0 0 dbfs:/databricks-datasets/songs/ songs/ 0 0 dbfs:/databricks-datasets/structured-streaming/ structured-streaming/ 0 0 dbfs:/databricks-datasets/timeseries/ timeseries/ 0 0 dbfs:/databricks-datasets/tpch/ tpch/ 0 0 dbfs:/databricks-datasets/warmup/ warmup/ 0 0 dbfs:/databricks-datasets/weather/ weather/ 0 0 dbfs:/databricks-datasets/wiki/ wiki/ 0 0 dbfs:/databricks-datasets/wikipedia-datasets/ wikipedia-datasets/ 0 0 dbfs:/databricks-datasets/wine-quality/ wine-quality/ 0 0

**spark is a spark session object**, and every spark program starts with spark session object.

In [0]:
spark

SparkSession - hive 
 
 
 SparkContext 

 Spark UI 

 
 Version 
 v3.2.1 
 Master 
 local[8] 
 AppName 
 Databricks Shell

#### Spark Dataframe 
2. **read** is an attribute of spark session object and it returns dataframe using format, option, load
3. **format**: mentioning to dataframe reader that dataframe is a csv file.
4. **option**: also to reader, csv file is comming with header row and it should follow inferShema
5. **load**: reader will read the data from the location.

In [0]:
fire_df = spark.read \
            .format("csv") \
            .option("header", "true") \
            .option("inferSchema", "true") \
            .load("/databricks-datasets/learning-spark-v2/sf-fire/sf-fire-calls.csv")


+-----------+-------+---------------+----------------+----------+----------+----------------------+--------------------+--------------------+----+-------------------+---------+------------+----+------------+--------+--------------+--------+---------------+---------+------------+------------------------------+------------------------+-------------------+--------------------+--------------------+-------------+------------------+
|Call Number|Unit ID|Incident Number|        CallType| Call Date|Watch Date|Call Final Disposition|      Available DtTm|             Address|City|Zipcode of Incident|Battalion|Station Area| Box|OrigPriority|Priority|Final Priority|ALS Unit|Call Type Group|NumAlarms|    UnitType|Unit sequence in call dispatch|Fire Prevention District|Supervisor District|        Neighborhood|            Location|        RowID|             Delay|
+-----------+-------+---------------+----------------+----------+----------+----------------------+--------------------+------------------

##### csv is a shortcut method, but prefferd to use above long form to read any type of file as its standardized

In [0]:
fire_df = spark.read \
            .csv("/databricks-datasets/learning-spark-v2/sf-fire/sf-fire-calls.csv", 
                header=True,
                inferSchema=True)

##### show() is not retuning dataframe nice way,

In [0]:
fire_df.show(10)

+-----------+-------+---------------+----------------+----------+----------+----------------------+--------------------+--------------------+----+-------------------+---------+------------+----+------------+--------+--------------+--------+---------------+---------+------------+------------------------------+------------------------+-------------------+---------------+--------------------+-------------+------------------+
|Call Number|Unit ID|Incident Number|        CallType| Call Date|Watch Date|Call Final Disposition|      Available DtTm|             Address|City|Zipcode of Incident|Battalion|Station Area| Box|OrigPriority|Priority|Final Priority|ALS Unit|Call Type Group|NumAlarms|    UnitType|Unit sequence in call dispatch|Fire Prevention District|Supervisor District|   Neighborhood|            Location|        RowID|             Delay|
+-----------+-------+---------------+----------------+----------+----------+----------------------+--------------------+--------------------+----+--

##### display() is showing dataframe in a tabular format. 
##### but its not belongs to a apache spark. Its a optional tool of databricks.

In [0]:
display(fire_df)

Call Number Unit ID Incident Number CallType Call Date Watch Date Call Final Disposition Available DtTm Address City Zipcode of Incident Battalion Station Area Box OrigPriority Priority Final Priority ALS Unit Call Type Group NumAlarms UnitType Unit sequence in call dispatch Fire Prevention District Supervisor District Neighborhood Location RowID Delay 20110014 M29 2003234 Medical Incident 01/11/2002 01/10/2002 Other 01/11/2002 01:58:43 AM 10TH ST/MARKET ST SF 94103 B02 36 2338 1 1 2 true null 1 MEDIC 1 2 6 Tenderloin (37.7765408927183, -122.417501464907) 020110014-M29 5.233333333333333 20110015 M08 2003233 Medical Incident 01/11/2002 01/10/2002 Other 01/11/2002 02:10:17 AM 300 Block of 5TH ST SF 94107 B03 08 2243 1 1 2 true null 1 MEDIC 1 3 6 South of Market (37.7792841462441, -122.402061300134) 020110015-M08 3.0833333333333335 20110016 B02 2003235 Structure Fire 01/11/2002 01/10/2002 Other 01/11/2002 01:47:00 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 CHIEF 6 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-B02 3.05 20110016 B04 2003235 Structure Fire 01/11/2002 01/10/2002 Other 01/11/2002 01:51:54 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 CHIEF 3 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-B04 2.316666666666667 20110016 D2 2003235 Structure Fire 01/11/2002 01/10/2002 Other 01/11/2002 01:47:00 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 CHIEF 4 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-D2 3.0166666666666666 20110016 E03 2003235 Structure Fire 01/11/2002 01/10/2002 Other 01/11/2002 01:47:00 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 ENGINE 7 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-E03 2.683333333333333 20110016 E38 2003235 Structure Fire 01/11/2002 01/10/2002 Other 01/11/2002 01:51:17 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 ENGINE 1 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-E38 2.1 20110016 E41 2003235 Structure Fire 01/11/2002 01/10/2002 Other 01/11/2002 01:47:00 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 ENGINE 8 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-E41 2.716666666666667 20110016 M03 2003235 Structure Fire 01/11/2002 01/10/2002 Other 01/11/2002 01:46:38 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 true null 1 MEDIC 10 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-M03 2.7666666666666666 20110016 RS1 2003235 Structure Fire 01/11/2002 01/10/2002 Other 01/11/2002 01:46:57 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 RESCUE SQUAD 9 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-RS1 3.2666666666666666 20110016 T03 2003235 Structure Fire 01/11/2002 01/10/2002 Other 01/11/2002 01:47:00 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 TRUCK 5 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-T03 3.2333333333333334 20110016 T13 2003235 Structure Fire 01/11/2002 01/10/2002 Other 01/11/2002 01:51:44 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 TRUCK 2 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-T13 2.95 20110017 M13 2003236 Medical Incident 01/11/2002 01/10/2002 Other 01/11/2002 02:27:14 AM 700 Block of FOLSOM ST SF 94107 B03 01 2176 1 1 2 true null 1 MEDIC 1 3 6 South of Market (37.7831757914875, -122.399603541822) 020110017-M13 6.933333333333334 20110019 M36 2003238 Medical Incident 01/11/2002 01/10/2002 Other 01/11/2002 02:28:30 AM 9TH ST/LARKIN ST SF 94103 B02 36 2337 1 1 2 true null 1 MEDIC 1 2 6 South of Market (37.7774938436091, -122.41629170277) 020110019-M36 6.116666666666666 20110020 E38 2003239 Outside Fire 01/11/2002 01/10/2002 Other 01/11/2002 02:15:33 AM 3000 Block of WEBSTER ST SF 94123 B04 16 3462 3 3 3 false null 1 ENGINE 1 4 2 Mar

In [0]:
fire_df.columns

Out[7]: ['Call Number',
 'Unit ID',
 'Incident Number',
 'CallType',
 'Call Date',
 'Watch Date',
 'Call Final Disposition',
 'Available DtTm',
 'Address',
 'City',
 'Zipcode of Incident',
 'Battalion',
 'Station Area',
 'Box',
 'OrigPriority',
 'Priority',
 'Final Priority',
 'ALS Unit',
 'Call Type Group',
 'NumAlarms',
 'UnitType',
 'Unit sequence in call dispatch',
 'Fire Prevention District',
 'Supervisor District',
 'Neighborhood',
 'Location',
 'RowID',
 'Delay']

In [0]:
fire_df.dtypes

Out[11]: [('Call Number', 'int'),
 ('Unit ID', 'string'),
 ('Incident Number', 'int'),
 ('CallType', 'string'),
 ('Call Date', 'string'),
 ('Watch Date', 'string'),
 ('Call Final Disposition', 'string'),
 ('Available DtTm', 'string'),
 ('Address', 'string'),
 ('City', 'string'),
 ('Zipcode of Incident', 'int'),
 ('Battalion', 'string'),
 ('Station Area', 'string'),
 ('Box', 'string'),
 ('OrigPriority', 'string'),
 ('Priority', 'string'),
 ('Final Priority', 'int'),
 ('ALS Unit', 'boolean'),
 ('Call Type Group', 'string'),
 ('NumAlarms', 'int'),
 ('UnitType', 'string'),
 ('Unit sequence in call dispatch', 'int'),
 ('Fire Prevention District', 'string'),
 ('Supervisor District', 'string'),
 ('Neighborhood', 'string'),
 ('Location', 'string'),
 ('RowID', 'string'),
 ('Delay', 'double')]

##### createGlobalTempView
1. Creating global temporary view similar like a table.
2. we can query through sql in this view

In [0]:
fire_df.createGlobalTempView("fire_service_calls_view")

##### global_temp is a hidden database
1. all global temporary tables, views are created inside it.

In [0]:
%sql

select * from global_temp.fire_service_calls_view

Call Number Unit ID Incident Number CallType Call Date Watch Date Call Final Disposition Available DtTm Address City Zipcode of Incident Battalion Station Area Box OrigPriority Priority Final Priority ALS Unit Call Type Group NumAlarms UnitType Unit sequence in call dispatch Fire Prevention District Supervisor District Neighborhood Location RowID Delay 20110014 M29 2003234 Medical Incident 01/11/2002 01/10/2002 Other 01/11/2002 01:58:43 AM 10TH ST/MARKET ST SF 94103 B02 36 2338 1 1 2 true null 1 MEDIC 1 2 6 Tenderloin (37.7765408927183, -122.417501464907) 020110014-M29 5.233333333333333 20110015 M08 2003233 Medical Incident 01/11/2002 01/10/2002 Other 01/11/2002 02:10:17 AM 300 Block of 5TH ST SF 94107 B03 08 2243 1 1 2 true null 1 MEDIC 1 3 6 South of Market (37.7792841462441, -122.402061300134) 020110015-M08 3.0833333333333335 20110016 B02 2003235 Structure Fire 01/11/2002 01/10/2002 Other 01/11/2002 01:47:00 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 CHIEF 6 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-B02 3.05 20110016 B04 2003235 Structure Fire 01/11/2002 01/10/2002 Other 01/11/2002 01:51:54 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 CHIEF 3 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-B04 2.316666666666667 20110016 D2 2003235 Structure Fire 01/11/2002 01/10/2002 Other 01/11/2002 01:47:00 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 CHIEF 4 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-D2 3.0166666666666666 20110016 E03 2003235 Structure Fire 01/11/2002 01/10/2002 Other 01/11/2002 01:47:00 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 ENGINE 7 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-E03 2.683333333333333 20110016 E38 2003235 Structure Fire 01/11/2002 01/10/2002 Other 01/11/2002 01:51:17 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 ENGINE 1 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-E38 2.1 20110016 E41 2003235 Structure Fire 01/11/2002 01/10/2002 Other 01/11/2002 01:47:00 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 ENGINE 8 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-E41 2.716666666666667 20110016 M03 2003235 Structure Fire 01/11/2002 01/10/2002 Other 01/11/2002 01:46:38 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 true null 1 MEDIC 10 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-M03 2.7666666666666666 20110016 RS1 2003235 Structure Fire 01/11/2002 01/10/2002 Other 01/11/2002 01:46:57 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 RESCUE SQUAD 9 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-RS1 3.2666666666666666 20110016 T03 2003235 Structure Fire 01/11/2002 01/10/2002 Other 01/11/2002 01:47:00 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 TRUCK 5 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-T03 3.2333333333333334 20110016 T13 2003235 Structure Fire 01/11/2002 01/10/2002 Other 01/11/2002 01:51:44 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 TRUCK 2 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-T13 2.95 20110017 M13 2003236 Medical Incident 01/11/2002 01/10/2002 Other 01/11/2002 02:27:14 AM 700 Block of FOLSOM ST SF 94107 B03 01 2176 1 1 2 true null 1 MEDIC 1 3 6 South of Market (37.7831757914875, -122.399603541822) 020110017-M13 6.933333333333334 20110019 M36 2003238 Medical Incident 01/11/2002 01/10/2002 Other 01/11/2002 02:28:30 AM 9TH ST/LARKIN ST SF 94103 B02 36 2337 1 1 2 true null 1 MEDIC 1 2 6 South of Market (37.7774938436091, -122.41629170277) 020110019-M36 6.116666666666666 20110020 E38 2003239 Outside Fire 01/11/2002 01/10/2002 Other 01/11/2002 02:15:33 AM 3000 Block of WEBSTER ST SF 94123 B04 16 3462 3 3 3 false null 1 ENGINE 1 4 2 Mar

## Dataframe Transformations

##### Rename Spark Dataframe Column

In [0]:
list_columns = fire_df.columns
for col in list_columns:
    print(f'.withColumnRenamed("{col}", "{col.replace(" ","")}")  \\')       

.withColumnRenamed("Call Number", "CallNumber")  \
.withColumnRenamed("Unit ID", "UnitID")  \
.withColumnRenamed("Incident Number", "IncidentNumber")  \
.withColumnRenamed("CallType", "CallType")  \
.withColumnRenamed("Call Date", "CallDate")  \
.withColumnRenamed("Watch Date", "WatchDate")  \
.withColumnRenamed("Call Final Disposition", "CallFinalDisposition")  \
.withColumnRenamed("Available DtTm", "AvailableDtTm")  \
.withColumnRenamed("Address", "Address")  \
.withColumnRenamed("City", "City")  \
.withColumnRenamed("Zipcode of Incident", "ZipcodeofIncident")  \
.withColumnRenamed("Battalion", "Battalion")  \
.withColumnRenamed("Station Area", "StationArea")  \
.withColumnRenamed("Box", "Box")  \
.withColumnRenamed("OrigPriority", "OrigPriority")  \
.withColumnRenamed("Priority", "Priority")  \
.withColumnRenamed("Final Priority", "FinalPriority")  \
.withColumnRenamed("ALS Unit", "ALSUnit")  \
.withColumnRenamed("Call Type Group", "CallTypeGroup")  \
.withColumnRenamed("NumAlarms",

In [0]:
renamed_fire_df = fire_df \
                    .withColumnRenamed("Call Number", "CallNumber")  \
                    .withColumnRenamed("Unit ID", "UnitID")  \
                    .withColumnRenamed("Incident Number", "IncidentNumber")  \
                    .withColumnRenamed("CallType", "CallType")  \
                    .withColumnRenamed("Call Date", "CallDate")  \
                    .withColumnRenamed("Watch Date", "WatchDate")  \
                    .withColumnRenamed("Call Final Disposition", "CallFinalDisposition")  \
                    .withColumnRenamed("Available DtTm", "AvailableDtTm")  \
                    .withColumnRenamed("Address", "Address")  \
                    .withColumnRenamed("City", "City")  \
                    .withColumnRenamed("Zipcode of Incident", "ZipcodeofIncident")  \
                    .withColumnRenamed("Battalion", "Battalion")  \
                    .withColumnRenamed("Station Area", "StationArea")  \
                    .withColumnRenamed("Box", "Box")  \
                    .withColumnRenamed("OrigPriority", "OrigPriority")  \
                    .withColumnRenamed("Priority", "Priority")  \
                    .withColumnRenamed("Final Priority", "FinalPriority")  \
                    .withColumnRenamed("ALS Unit", "ALSUnit")  \
                    .withColumnRenamed("Call Type Group", "CallTypeGroup")  \
                    .withColumnRenamed("NumAlarms", "NumAlarms")  \
                    .withColumnRenamed("UnitType", "UnitType")  \
                    .withColumnRenamed("Unit sequence in call dispatch", "Unitsequenceincalldispatch")  \
                    .withColumnRenamed("Fire Prevention District", "FirePreventionDistrict")  \
                    .withColumnRenamed("Supervisor District", "SupervisorDistrict")  \
                    .withColumnRenamed("Neighborhood", "Neighborhood")  \
                    .withColumnRenamed("Location", "Location")  \
                    .withColumnRenamed("RowID", "RowID")  \
                    .withColumnRenamed("Delay", "Delay")

##### Check the Spark Dataframe columns and its datatype

In [0]:
renamed_fire_df.dtypes

Out[34]: [('CallNumber', 'int'),
 ('UnitID', 'string'),
 ('IncidentNumber', 'int'),
 ('CallType', 'string'),
 ('CallDate', 'string'),
 ('WatchDate', 'string'),
 ('CallFinalDisposition', 'string'),
 ('AvailableDtTm', 'string'),
 ('Address', 'string'),
 ('City', 'string'),
 ('ZipcodeofIncident', 'int'),
 ('Battalion', 'string'),
 ('StationArea', 'string'),
 ('Box', 'string'),
 ('OrigPriority', 'string'),
 ('Priority', 'string'),
 ('FinalPriority', 'int'),
 ('ALSUnit', 'boolean'),
 ('CallTypeGroup', 'string'),
 ('NumAlarms', 'int'),
 ('UnitType', 'string'),
 ('Unitsequenceincalldispatch', 'int'),
 ('FirePreventionDistrict', 'string'),
 ('SupervisorDistrict', 'string'),
 ('Neighborhood', 'string'),
 ('Location', 'string'),
 ('RowID', 'string'),
 ('Delay', 'double')]

In [0]:
renamed_fire_df.printSchema()

root
 |-- CallNumber: integer (nullable = true)
 |-- UnitID: string (nullable = true)
 |-- IncidentNumber: integer (nullable = true)
 |-- CallType: string (nullable = true)
 |-- CallDate: string (nullable = true)
 |-- WatchDate: string (nullable = true)
 |-- CallFinalDisposition: string (nullable = true)
 |-- AvailableDtTm: string (nullable = true)
 |-- Address: string (nullable = true)
 |-- City: string (nullable = true)
 |-- ZipcodeofIncident: integer (nullable = true)
 |-- Battalion: string (nullable = true)
 |-- StationArea: string (nullable = true)
 |-- Box: string (nullable = true)
 |-- OrigPriority: string (nullable = true)
 |-- Priority: string (nullable = true)
 |-- FinalPriority: integer (nullable = true)
 |-- ALSUnit: boolean (nullable = true)
 |-- CallTypeGroup: string (nullable = true)
 |-- NumAlarms: integer (nullable = true)
 |-- UnitType: string (nullable = true)
 |-- Unitsequenceincalldispatch: integer (nullable = true)
 |-- FirePreventionDistrict: string (nullable = t

In [0]:
from pyspark.sql.functions import to_date, to_timestamp, round

fire_df = renamed_fire_df \
            .withColumn("CallDate", to_date("CallDate", "MM/dd/yyyy"))  \
            .withColumn("WatchDate", to_date("WatchDate", "MM/dd/yyyy"))  \
            .withColumn("AvailableDtTm", to_timestamp("AvailableDtTm", "MM/dd/yyyy hh:mm:ss a"))  \
            .withColumn("Delay", round("Delay", 2))

In [0]:
display(fire_df)

CallNumber UnitID IncidentNumber CallType CallDate WatchDate CallFinalDisposition AvailableDtTm Address City ZipcodeofIncident Battalion StationArea Box OrigPriority Priority FinalPriority ALSUnit CallTypeGroup NumAlarms UnitType Unitsequenceincalldispatch FirePreventionDistrict SupervisorDistrict Neighborhood Location RowID Delay 20110014 M29 2003234 Medical Incident 2002-01-11 2002-01-10 Other 2002-01-11T01:58:43.000+0000 10TH ST/MARKET ST SF 94103 B02 36 2338 1 1 2 true null 1 MEDIC 1 2 6 Tenderloin (37.7765408927183, -122.417501464907) 020110014-M29 5.23 20110015 M08 2003233 Medical Incident 2002-01-11 2002-01-10 Other 2002-01-11T02:10:17.000+0000 300 Block of 5TH ST SF 94107 B03 08 2243 1 1 2 true null 1 MEDIC 1 3 6 South of Market (37.7792841462441, -122.402061300134) 020110015-M08 3.08 20110016 B02 2003235 Structure Fire 2002-01-11 2002-01-10 Other 2002-01-11T01:47:00.000+0000 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 CHIEF 6 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-B02 3.05 20110016 B04 2003235 Structure Fire 2002-01-11 2002-01-10 Other 2002-01-11T01:51:54.000+0000 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 CHIEF 3 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-B04 2.32 20110016 D2 2003235 Structure Fire 2002-01-11 2002-01-10 Other 2002-01-11T01:47:00.000+0000 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 CHIEF 4 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-D2 3.02 20110016 E03 2003235 Structure Fire 2002-01-11 2002-01-10 Other 2002-01-11T01:47:00.000+0000 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 ENGINE 7 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-E03 2.68 20110016 E38 2003235 Structure Fire 2002-01-11 2002-01-10 Other 2002-01-11T01:51:17.000+0000 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 ENGINE 1 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-E38 2.1 20110016 E41 2003235 Structure Fire 2002-01-11 2002-01-10 Other 2002-01-11T01:47:00.000+0000 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 ENGINE 8 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-E41 2.72 20110016 M03 2003235 Structure Fire 2002-01-11 2002-01-10 Other 2002-01-11T01:46:38.000+0000 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 true null 1 MEDIC 10 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-M03 2.77 20110016 RS1 2003235 Structure Fire 2002-01-11 2002-01-10 Other 2002-01-11T01:46:57.000+0000 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 RESCUE SQUAD 9 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-RS1 3.27 20110016 T03 2003235 Structure Fire 2002-01-11 2002-01-10 Other 2002-01-11T01:47:00.000+0000 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 TRUCK 5 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-T03 3.23 20110016 T13 2003235 Structure Fire 2002-01-11 2002-01-10 Other 2002-01-11T01:51:44.000+0000 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 TRUCK 2 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-T13 2.95 20110017 M13 2003236 Medical Incident 2002-01-11 2002-01-10 Other 2002-01-11T02:27:14.000+0000 700 Block of FOLSOM ST SF 94107 B03 01 2176 1 1 2 true null 1 MEDIC 1 3 6 South of Market (37.7831757914875, -122.399603541822) 020110017-M13 6.93 20110019 M36 2003238 Medical Incident 2002-01-11 2002-01-10 Other 2002-01-11T02:28:30.000+0000 9TH ST/LARKIN ST SF 94103 B02 36 2337 1 1 2 true null 1 MEDIC 1 2 6 South of Market (37.7774938436091, -122.41629170277) 020110019-M36 6.12 20110020 E38 2003239 Outside Fire 2002-01-11 2002-01-10 Other 2002-01-11T02:15:33.000+0000 3000 Block of WEBSTER ST SF 94123 B04 16 3462 3 3 3 false null 1 ENGINE 1 4 2 Marina (37.7984543611006, -122.434151287383) 020110020-E38 1.48 20110021 E03 200324

In [0]:
fire_df.printSchema()

root
 |-- CallNumber: integer (nullable = true)
 |-- UnitID: string (nullable = true)
 |-- IncidentNumber: integer (nullable = true)
 |-- CallType: string (nullable = true)
 |-- CallDate: date (nullable = true)
 |-- WatchDate: date (nullable = true)
 |-- CallFinalDisposition: string (nullable = true)
 |-- AvailableDtTm: timestamp (nullable = true)
 |-- Address: string (nullable = true)
 |-- City: string (nullable = true)
 |-- ZipcodeofIncident: integer (nullable = true)
 |-- Battalion: string (nullable = true)
 |-- StationArea: string (nullable = true)
 |-- Box: string (nullable = true)
 |-- OrigPriority: string (nullable = true)
 |-- Priority: string (nullable = true)
 |-- FinalPriority: integer (nullable = true)
 |-- ALSUnit: boolean (nullable = true)
 |-- CallTypeGroup: string (nullable = true)
 |-- NumAlarms: integer (nullable = true)
 |-- UnitType: string (nullable = true)
 |-- Unitsequenceincalldispatch: integer (nullable = true)
 |-- FirePreventionDistrict: string (nullable = tr

## Data Analysis on Fire Calls Dataframe

#### Cache the Spark Dataframe

In [0]:
fire_df.cache()

Out[43]: DataFrame[CallNumber: int, UnitID: string, IncidentNumber: int, CallType: string, CallDate: date, WatchDate: date, CallFinalDisposition: string, AvailableDtTm: timestamp, Address: string, City: string, ZipcodeofIncident: int, Battalion: string, StationArea: string, Box: string, OrigPriority: string, Priority: string, FinalPriority: int, ALSUnit: boolean, CallTypeGroup: string, NumAlarms: int, UnitType: string, Unitsequenceincalldispatch: int, FirePreventionDistrict: string, SupervisorDistrict: string, Neighborhood: string, Location: string, RowID: string, Delay: double]

##### Q1. How many distinct types of calls were made to the Fire Department?
```sql
select count(distinct CallType) as distinct_call_type_count 
from demo_db.fire_service_calls_tbl
where CallType is not null
```

In [0]:
fire_df.createOrReplaceTempView("fire_service_calls_view")
q1_sql_df = spark.sql("""
            select count(distinct CallType) as distinct_call_type_count 
            from fire_service_calls_view
            where CallType is not null
        """)
display(q1_sql_df)

distinct_call_type_count 32

In [0]:
q1_df = fire_df.where("CallType is not null")  \
                .select("CallType")  \
                .distinct()
print(q1_df.count())

32


##### Q2. What are the distinct types of calls made to the fire department?
```sql
select distinct CallType as distinct_call_type 
from demo_db.fire_service_calls_tbl
where CallType is not null
```

In [0]:
q2_df = fire_df.where("CallType is not null")  \
                .select(expr("CallType as distinct_CallType"))  \
                .distinct()
display(q2_df)

distinct_CallType Elevator / Escalator Rescue Marine Fire Aircraft Emergency Confined Space / Structure Collapse Administrative Alarms Odor (Strange / Unknown) Citizen Assist / Service Call HazMat Watercraft in Distress Explosion Oil Spill Vehicle Fire Suspicious Package Extrication / Entrapped (Machinery, Vehicle) Other Outside Fire Traffic Collision Assist Police Gas Leak (Natural and LP Gases) Water Rescue Electrical Hazard High Angle Rescue Structure Fire Industrial Accidents Medical Incident Mutual Aid / Assist Outside Agency Fuel Spill Smoke Investigation (Outside) Train / Rail Incident Train / Rail Fire Lightning Strike (Investigation)

##### Q3. Find out all responses for delayed times greater than 5 mints?
```sql
select CallNumber, Delay
from demo_db.fire_service_calls_tbl
where Delay > 5
```

In [0]:
q3_df = fire_df.where("Delay > 5")  \
                .select(["CallNumber", "Delay"])
display(q3_df)

CallNumber Delay 20110014 5.23 20110017 6.93 20110019 6.12 20110039 7.85 20110045 77.33 20110046 5.42 20110055 6.5 20110058 6.85 20110058 6.85 20110061 6.33 20110062 7.02 20110066 9.1 20110068 5.98 20110070 5.42 20110070 5.08 20110070 8.18 20110070 5.98 20110077 18.07 20110103 5.55 20110103 5.97 20110103 6.05 20110141 6.13 20110148 6.27 20110157 6.3 20110162 5.63 20110163 5.05 20110163 5.63 20110170 12.73 20110180 6.15 20110180 7.3 20110190 5.38 20110196 8.0 20110196 7.77 20110198 5.25 20110237 5.05 20110240 7.52 20110245 5.92 20110252 5.2 20110268 21.12 20110271 5.22 20110273 5.4 20110276 8.7 20110277 8.22 20110277 8.45 20110289 10.3 20110297 6.5 20110310 5.03 20110315 5.35 20110334 5.1 20110336 11.07 20110338 28.25 20110338 6.45 20110338 6.3 20110338 7.25 20110338 6.25 20110342 7.0 20110348 8.65 20110349 7.32 20110357 5.07 20110360 6.55 20110367 11.38 20110367 10.53 20110367 10.57 20110368 5.67 20110369 8.37 20110378 12.52 20110383 5.32 20110388 8.55 20110389 16.72 20110390 5.35 20110400 12.7 20110405 6.0 20110410 8.28 20110421 10.88 20110421 10.57 20110424 7.7 20110424 5.07 20110424 5.2 20110427 8.63 20110429 6.65 20110435 6.07 20110437 7.53 20110455 20.37 20110469 5.52 20110473 7.33 20110475 5.18 20120026 7.27 20120027 5.3 20120031 5.55 20120034 5.92 20120038 17.02 20120039 9.1 20120040 6.03 20120047 10.2 20120059 9.12 20120059 8.58 20120076 5.65 20120104 7.27 20120108 6.57 20120137 11.72 20120147 6.25 20120153 7.45 20120155 5.33 20120156 14.67 20120156 8.95 20120163 7.05 20120164 5.52 20120173 6.65 20120173 9.6 20120173 8.35 20120173 6.95 20120191 5.67 20120195 6.5 20120206 5.17 20120208 7.4 20120213 5.63 20120230 7.57 20120233 5.05 20120235 41.37 20120247 6.13 20120260 5.68 20120268 5.18 20120272 8.17 20120284 5.07 20120296 5.07 20120312 6.2 20120334 5.3 20120339 5.27 20120342 5.03 20120349 6.23 20130013 5.2 20130020 20.57 20130041 5.42 20130041 5.73 20130042 6.17 20130042 5.42 20130047 5.58 20130053 13.82 20130054 5.55 20130055 8.0 20130063 9.28 20130065 9.2 20130068 8.43 20130077 6.72 20130094 7.47 20130109 7.32 20130122 6.35 20130131 6.25 20130136 10.1 20130140 5.03 20130153 5.58 20130159 5.65 20130165 5.23 20130170 6.32 20130178 5.62 20130180 8.68 20130182 5.32 20130183 5.3 20130187 7.77 20130195 7.43 20130198 20.98 20130202 6.1 20130216 12.23 20130219 5.3 20130220 5.22 20130224 9.1 20130225 8.35 20130230 8.43 20130247 17.22 20130260 6.2 20130260 6.3 20130268 5.83 20130268 5.08 20130272 8.13 20130274 8.35 20130274 8.35 20130274 8.35 20130290 6.48 20130291 7.97 20130293 5.95 20130295 11.7 20130297 5.38 20130301 5.6 20130301 6.3 20130318 5.22 20130318 7.58 20130319 10.38 20130319 20.8 20130355 5.75 20140005 5.88 20140020 8.23 20140022 5.48 20140026 5.65 20140033 5.18 20140034 6.67 20140035 5.6 20140039 7.2 20140045 6.53 20140046 5.43 20140053 5.22 20140062 8.02 20140067 5.6 20140074 6.43 20140074 5.23 20140102 22.65 20140102 22.65 20140106 5.15 20140128 6.5 20140128 5.18 20140130 5.78 20140137 6.75 20140139 5.2 20140139 6.27 20140142 10.08 20140150 5.45 20140158 9.67 20140167 6.28 20140177 7.25 20140196 6.25 20140202 7.23 20140215 10.0 20140215 9.73 20140215 10.32 20140221 5.05 20140235 8.12 20140247 6.23 20140260 15.47 20140265 5.52 20140266 6.82 20140288 6.23 20140289 7.1 20140292 10.73 20140294 5.08 20140305 5.3 20140317 11.2 20140324 5.52 20140358 5.43 20140373 6.9 20150003 5.8 20150003 5.12 20150010 6.87 20150012 6.1 20150012 5.42 20150030 5.02 20150036 5.27 20150050 7.37 20150050 5.1 20150056 11.92 20150075 6.93 20150118 9.8 20150120 6.45 20150146 8.57 20150146 5.32 20150162 16.13 20150168 7.22 20150206 6.37 20150206 6.25 20150215 5.23 20150217 5.67 20150218 18.0 20150227 5.28 20150227 5.23 20150228 6.53 20150229 8.25 20150229 8.03 20150229 7.93 20150234 8.78 20150234 9.35 20150247 5.43 20150252 6.85 20150252 8.27 20150253 12.43 20150254 5.12 20150258 8.48 20150265 10.07 20150265 8.63 20150265 95.28 20150267 5.93 20150270 8.22 20150270 8.27 20150270 8.22 20150271 7.5 20150272 5.1 20150272 7.08 20150281 5.32 20150

##### Q4. What were the most common call types?
```sql
select CallType, count(*) as count
from demo_db.fire_service_calls_tbl
where CallType is not null
group by CallType
order by count desc
```

1. count is action or transformation?
    1. DataFrame.count() -> Action
    2. GroupedData.count() -> Transformation

In [0]:
fire_df.select("CallType")  \
    .where("CallType is not null")  \
    .groupBy("CallType") \
    .count() \
    .orderBy("count", ascending=False) \
    .show()
    

+--------------------+-------+
|            CallType|  count|
+--------------------+-------+
|    Medical Incident|2843475|
|      Structure Fire| 578998|
|              Alarms| 483518|
|   Traffic Collision| 175507|
|Citizen Assist / ...|  65360|
|               Other|  56961|
|        Outside Fire|  51603|
|        Vehicle Fire|  20939|
|        Water Rescue|  20037|
|Gas Leak (Natural...|  17284|
|   Electrical Hazard|  12608|
|Elevator / Escala...|  11851|
|Odor (Strange / U...|  11680|
|Smoke Investigati...|   9796|
|          Fuel Spill|   5198|
|              HazMat|   3437|
|Industrial Accidents|   2401|
|           Explosion|   2307|
|  Aircraft Emergency|   1196|
|Train / Rail Inci...|   1116|
+--------------------+-------+
only showing top 20 rows



##### Q5. What zip codes accounted for the most common calls?
```sql
select CallType, zipcode, count(*) as count
from demo_db.fire_service_calls_tbl
where CallType is not null
group by CallType, zipcode
order by count desc
```

##### Q6. What San Francisco neighborhoods are in zip codes 94102 and 94103
```sql
select zipcode, Neighborhood
from demo_db.fire_service_calls_tbl
where zipcode == 94102 or zipcode == 94103
```

##### Q7. What was the sum of all call alarms, average, min and max of the call response times?
```sql
select Sum(NumAlarms) as sum_call_alarms, 
avg(delay) as avg_delay, 
min(delay) as min_delay, 
max(delay) as max_delay
from demo_db.fire_service_calls_tbl
```

##### Q8. How many distinct years of data are in csv files?
```sql
select distinct year(to_date(CallDate, "MM/dd/yyyy")) as distinct_year
from demo_db.fire_service_calls_tbl
```

##### Q9. What week of the year in 2018 had the most fire calls?
```sql
select weekofyear(to_date(CallDate, "MM/dd/yyyy")) as week_year, count(*) as count 
from demo_db.fire_service_calls_tbl
where year(to_date(CallDate, "MM/dd/yyyy")) = 2018
group by week_year
order by count desc
```

##### Q10. What neighborhoods in San Francisco had the worst response time in 2018?
```sql
select Neighborhood, delay
from demo_db.fire_service_calls_tbl
where year(to_date(CallDate, "MM/dd/yyyy")) = 2018
order by delay desc
```